File: src/infinicore/ops/syr2k/syr2k.cc
#include "infinicore/ops/syr2k.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(Syr2k);

Syr2k::Syr2k(const Tensor &a, const Tensor &b, const Tensor &alpha, const Tensor &beta, Tensor c, int uplo, int trans) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(a, b, alpha, beta, c);
    INFINICORE_GRAPH_OP_DISPATCH(c->device().getType(), a, b, alpha, beta, c, uplo, trans);
}

void Syr2k::execute(const Tensor &a, const Tensor &b, const Tensor &alpha, const Tensor &beta, Tensor c, int uplo, int trans) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(Syr2k, a, b, alpha, beta, c, uplo, trans);
}

void syr2k_(const Tensor &a, const Tensor &b, const Tensor &alpha, const Tensor &beta, Tensor c, int uplo, int trans) {
    Syr2k::execute(a, b, alpha, beta, c, uplo, trans);
}

} // namespace infinicore::op

File: python/infinicore/ops/syr2k.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def syr2k(
    a: Tensor,
    b: Tensor,
    alpha: Tensor,
    beta: Tensor,
    out: Tensor,
    *,
    uplo: int = 0,
    trans: int = 0,
):
    _infinicore.syr2k_(
        a._underlying,
        b._underlying,
        alpha._underlying,
        beta._underlying,
        out._underlying,
        uplo,
        trans,
    )

    return out
